blib/man3/JavaArray.3
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH JavaArray 3 "perl 5.005, patch 03" "12/Mar/2001" "User Contributed Perl Documentation"
.UC
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
JavaArray \- Tie'd extension for Java arrays from Java.pm
.SH "SYNOPSIS"
.PP
.Vb 1
\&  use Java;
.Ve
.Vb 3
\&  # Set up Java.pm to always return tied array references to me
\&  my $java = new Java(use_tied_arrays => 1);
\&  my $tied_array = $java->create_array("java.lang.String",5);
.Ve
.Vb 1
\&        OR
.Ve
.Vb 5
\& # Roll my own tied arrays
\&        my @tied_array;
\&        tie @tied_array, 'JavaArray', $java->create_array("java.lang.String",5);
\&                OR
\&        tie @tied_array, 'JavaArray', $some_object_that_is_an_array;
.Ve
.Vb 2
\&  // Set array element 3 to "Java is lame"
\&  $tied_array[3] = "Java is lame";
.Ve
.Vb 6
\&  // Get array element 3's value
\&  my $element = $tied_array[3]->get_value();
\&  
\&  // Get length
\&  my $length = scalar(@tied_array);
\&  my $size = $#tied_array;
.Ve
.Vb 2
\&  // Use as parameter you gotta pass the reference!
\&  my $list = $java->java_util_Arrays("asList",\e@tied_array);
.Ve
.Vb 4
\&  // NO OTHER ARRAY OPERATIONS ARE AVAILABLE!
\&  //    so no pop or push or unshift or shift or splice
\&  //    Hey event this can't make Java arrays cool!
\&  //    use the Collections framework!
.Ve
.SH "DESCRIPTION"
This module puts a pretty thin veneer over Java.pm objects are are Java
arrays.  Makes \*(L'em slightly prettier to play with.
You can pass as an agrument to the \*(L'tie\*(R' any Java object that is an array \-
either one you created yourself or one that was returned to you by
something else.
.PP
You probably should NOT be using this directly, but specify \*(L'use_tied_arrays\*(R'
in your constructor args to Java.pm.
.Sh "In parameter lists"
If you want to use your array in a parameter list you've got to pass in
the \s-1REFERENCE\s0 to your array or things will go haywire...
If you specified \*(L'use_tied_arrays\*(R' in your Java.pm constructor then you
will only receive references back from Java.pm so you've already got the
reference.  \s-1ONLY\s0 if you call \*(L'tie\*(R' yourself (& I can't really think of why
you ever would... \s-1BUT\s0) & get the array itself do you need to take its
reference when using it in parameter lists.
.Sh "Automatic usage"
You can tell Java.pm to automatically convert all Java arrays to their
tied counterparts by setting \*(L'use_tied_arrays\*(R' in your Java constructor.
You will get receive a reference to the tied array so you must use the \*(L'\->\*(R'
notation like:
.PP
.Vb 2
\&        $array->[3] = "Mark rox";
\&        my $ele = $array->[4];
.Ve
See perldoc Java.pm for more info.  You can then use that value directly
in parameter lists.
.Sh "\s-1EXPORT\s0"
None by default.
.SH "AUTHOR"
Mark Ethan Trostler, mark@zzo.com
.SH "SEE ALSO"
\fIperl\fR\|(1).
Java.pm.

.rn }` ''
.IX Title "JavaArray 3"
.IX Name "JavaArray - Tie'd extension for Java arrays from Java.pm"

.IX Header "NAME"

.IX Header "SYNOPSIS"

.IX Header "DESCRIPTION"

.IX Subsection "In parameter lists"

.IX Subsection "Automatic usage"

.IX Subsection "\s-1EXPORT\s0"

.IX Header "AUTHOR"

.IX Header "SEE ALSO"



Changes
Revision history for Perl extension Java.

1.0  Fri Apr 28 16:36:29 2000
	- original version; created by h2xs 1.20 with options
		-X -n Java -v 1.0
1.1  
	 - Allow creation of primitive Java array types
	 - Allow Strings to be created using encodings for Localization
		w/':string_<ENCODING>' syntax
	 - Allow integers to be strings w/':string' syntax
1.12
	- Refined string encoding stuff
1.13
	- Added 'static_call' function to make static method calls on
		classes NOT in packages.
1.14
	- Just setsockopt SO_REUSEADDR on Java.pm event_server port
2.0
	8/23/00
	- Now using ctrl-A to separate parameter lists so we can accept
		commas or any other printable character in strings.
		THIS CHANGE MEANS NEITHER JAVASERVER NOR JAVA.PM ARE
			COMPATIBLE WITH ANY 1.X RELEASE!!
		(hence the bump up of the major version number!)

Java stuff got updated to 2.1 - Perl stuff still at 2.0
        1/24/01
        - Included Andreas Leue's code to statically start and stop
                JavaServer from within an already-running Java app
                Check out 'test_test.pl' in the Perl examples directory
                so see what this looks like on the Perl side.  Pretty
                much the same!
                On the Java side you can now use 'JavaServer.start()',
                or 'JavaServer.start(int port)' to specify a port other
                than 2000 to programmatically start JavaServer & you
                similarly use 'JavaServer.stop()' to stop the madness
                when you've had enough fun.
                Like:

                public class MyApp
                {
                        public static void main(String args[])
                        {
                                // Allow JavaServer to 'hook' into
                                //      your running program
                                JavaServer.start();


                                // Do whatever yer program does

                                // Done
                                JavaServer.stop();
                        }
                }
        - Fixed stupid bug about the MANIFEST file looking for Java.jar
                when it's really called JavaServer.jar

Perl stuff up to 2.1 - Java stuff the same
	2/22/01
	- Fixed bug in 'static_call' so it now uses PARAMETER_SEPARATOR
		(thanks to David Roth)
	- Fixed bugs in java_server_install.pl to correctly reference 
		'JavaServer.jar' & not 'Java.jar'
		(thanks to Slavan Rezic & David Roth)
Perl stuff up to 2.2 - Java stuff the same
	3/9/01
	- Overloaded '==' to be the same as the 'same' function.
		So instead of having to do:
			if ($obj1.same($obj2))
		You can now do:
			if ($obj1 == $obj2)
		Check out the perldoc.
	- Added JavaArray tied interface for Java arrays!
		See 'perldoc Java' for all the glorious details!
		It's very cool!  See 'array.pl' in 'examples' directory
		for example usage.
	- Some other random bug fixes for the install stuff

Both Perl & Java stuff moved up to 3.0
	3/21/01
	- Moved up major version numbers because the JavaServer protocol
		has changed & is incompatible with previous versions.
		It was changed to handle multi-line sending.  Now all commands
		sent & received must be terminated by a Control-A beginning
		a line by itself (akin to SMTP and the '.' to end DATA).
	- So now you can say something like:
		my $str = $java->create_object("java.lang.String","multi\nline\string);
		As well as correctly fetch the value using 'get_value'.
	- Should ONLY effect multi-line String processing BUT since the
		protocol itself has changed yer gonna need 3.x versions so
		they can understand each other.
Perl stuff up to 3.1 Java stuff the same
	5/16/01
	- AUTOLOAD function now correctly recognizes if it's dealing with
		a static or instantiated method call - THEREFORE you can now
		call instantiated functions with '_'s in their names.
	- Also threw in some reorg of how functions get called to isolate
		the code.


blib/lib/JavaArray.pm
package JavaArray;

# Just keep track of java object
# A VERY (ridiculously so perhaps!) thin veneer over the stuff in Java.pm...
sub TIEARRAY
{
        my $class = shift;
        my $java_object = shift;

        return bless {
                java_object => $java_object,
        }, $class;
}

sub FETCH
{
        my($self,$index) = @_;
        $self->get_object->get_field($index);
}

sub STORE
{
        my($self,$index,$value) = @_;
        $self->get_object->set_field($index,$value);
}

sub FETCHSIZE
{
        my($self) = @_;
        $self->get_object->get_length;

}

sub STORESIZE
{
        my($self) = @_;
        $self->FETCHSIZE+1;
}

#sub DESTROY
#{
        #my($self) = @_;
        #$self->get_object->DESTROY;
#}

sub get_object
{
        my($self) = @_;
        $self->{java_object};
}

# Sneaky way to get actual java object outta this thang
#       for function calls & such w/o having to keep around
#       the actual object returned by 'tie'
sub POP
{
        my($self) = @_;
        $self->get_object;
}
1;
__END__

=head1 NAME

JavaArray - Tie'd extension for Java arrays from Java.pm

=head1 SYNOPSIS

  use Java;

  # Set up Java.pm to always return tied array references to me
  my $java = new Java(use_tied_arrays => 1);
  my $tied_array = $java->create_array("java.lang.String",5);

	OR

 # Roll my own tied arrays
	my @tied_array;
  	tie @tied_array, 'JavaArray', $java->create_array("java.lang.String",5);
        	OR
  	tie @tied_array, 'JavaArray', $some_object_that_is_an_array;

  // Set array element 3 to "Java is lame"
  $tied_array[3] = "Java is lame";

  // Get array element 3's value
  my $element = $tied_array[3]->get_value();
  
  // Get length
  my $length = scalar(@tied_array);
  my $size = $#tied_array;

  // Use as parameter you gotta pass the reference!
  my $list = $java->java_util_Arrays("asList",\@tied_array);

  // NO OTHER ARRAY OPERATIONS ARE AVAILABLE!
  //    so no pop or push or unshift or shift or splice
  //    Hey event this can't make Java arrays cool!
  //    use the Collections framework!

=head1 DESCRIPTION

This module puts a pretty thin veneer over Java.pm objects are are Java
arrays.  Makes 'em slightly prettier to play with.
You can pass as an agrument to the 'tie' any Java object that is an array -
either one you created yourself or one that was returned to you by
something else.

You probably should NOT be using this directly, but specify 'use_tied_arrays'
in your constructor args to Java.pm.

=head2 In parameter lists

If you want to use your array in a parameter list you've got to pass in
the REFERENCE to your array or things will go haywire...
If you specified 'use_tied_arrays' in your Java.pm constructor then you
will only receive references back from Java.pm so you've already got the
reference.  ONLY if you call 'tie' yourself (& I can't really think of why
you ever would... BUT) & get the array itself do you need to take its
reference when using it in parameter lists.

=head2 Automatic usage

You can tell Java.pm to automatically convert all Java arrays to their
tied counterparts by setting 'use_tied_arrays' in your Java constructor.
You will get receive a reference to the tied array so you must use the '->'
notation like:

        $array->[3] = "Mark rox";
        my $ele = $array->[4];

See perldoc Java.pm for more info.  You can then use that value directly
in parameter lists.

=head2 EXPORT

None by default.

=head1 AUTHOR

Mark Ethan Trostler, mark@zzo.com

=head1 SEE ALSO

perl(1).
Java.pm.

=cut
